google-cloud-firestore/lib/google/cloud/firestore/document_listener.rb

```diff
@@ -119,6 +126,83 @@ def stop
         def stopped?
           @listener.stopped?
         end
+
+        ##
+        # Register to be notified of errors when raised.
+        #
+        # If an unhandled error has occurred the listener will attempt to
+        # recover from the error and resume listening.
+        #
+        # Multiple error handlers can be added.
+        #
+        # @yield [callback] The block to be called when an error is raised.
+        # @yieldparam [Exception] error The error raised.
+        #
+        # @example
+        #   require "google/cloud/firestore"
+        #
+        #   firestore = Google::Cloud::Firestore.new
+        #
+        #   # Get a document reference
+        #   nyc_ref = firestore.doc "cities/NYC"
+        #
+        #   listener = nyc_ref.listen do |snapshot|
+        #     puts "The population of #{snapshot[:name]} "
+        #     puts "is #{snapshot[:population]}."
+        #   end
+        #
+        #   # Register to be notified when unhandled errors occur.
+        #   listener.on_error do |error|
+        #     # log error
+        #     puts error
+        #   end
+        #
+        #   # When ready, stop the listen operation and close the stream.
+        #   listener.stop
+        #
+        def on_error &block
+          synchronize { @error_callbacks << block }
```

Comment:
Might be a good idea to check for `nil` (i.e. missing block) explicitly and raise an exception here. Otherwise if someone does make this mistake (which is not hard to make), it wouldn't manifest until an actual error happens. Also see the similar place in QueryListener below.

Reply:
Correct, calling `on_error` without a block results in:

```
google-cloud-ruby/google-cloud-firestore/lib/google/cloud/firestore/query_listener.rb:193:in `block in error!': undefined method `call' for nil:NilClass (NoMethodError)
```

We will need to fix this in the other gems that have `on_error` methods.

Reply:
Fixed.